cmd/snap-update-ns/utils.go

```diff
@@ -286,6 +291,63 @@ func secureMkfileAll(name string, perm os.FileMode, uid, gid int) error {
 	return err
 }
 
+func planWritableMimic(dir string) ([]*Change, error) {
+	// We need a place for "safe keeping" of what is present in the original
+	// directory as we are about to attach a tmpfs there, which will hide
+	// everything inside.
+	logger.Debugf("create-writable-mimic %q", dir)
+	safeKeepingDir := filepath.Join("/tmp/.snap/", dir)
+
+	var changes []*Change
+
+	// Bind mount the original directory elsewhere for safe-keeping.
+	changes = append(changes, &Change{
+		Action: Mount, Entry: mount.Entry{
+			// XXX: should we rbind here?
+			Name: dir, Dir: safeKeepingDir, Options: []string{"bind"}},
+	})
+	// Mount tmpfs over the original directory, hiding its contents.
+	changes = append(changes, &Change{
+		Action: Mount, Entry: mount.Entry{Name: "none", Dir: dir, Type: "tmpfs"},
+	})
+	// Iterate over the items in the original directory (nothing is mounted _yet_).
+	entries, err := ioutilReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+	for _, fi := range entries {
+		ch := &Change{Action: Mount, Entry: mount.Entry{
+			Name: filepath.Join(safeKeepingDir, fi.Name()),
+			Dir:  filepath.Join(dir, fi.Name()),
+			// XXX: should we rbind here?
+			Options: []string{"bind", "ro"},
+		}}
+		// Bind mount each element from the safe-keeping directory into the
+		// tmpfs. Our Change.Perform() engine can create the missing
+		// directories automatically so we don't bother creating those.
+		m := fi.Mode()
+		switch {
+		case m.IsDir():
+			changes = append(changes, ch)
+		case m.IsRegular():
+			ch.Entry.Options = append(ch.Entry.Options, "x-snapd.kind=file")
+			changes = append(changes, ch)
+		case m&os.ModeSymlink != 0:
+			if target, err := osReadlink(filepath.Join(dir, fi.Name())); err == nil {
+				ch.Entry.Options = append(ch.Entry.Options, "x-snapd.kind=symlink", fmt.Sprintf("x-snapd.symlink=%s", target))
+				changes = append(changes, ch)
+			}
+		default:
+			logger.Noticef("skipping unsupported thing %s", fi)
```

Comment:
Maybe "skipping unsupported file %s"? Bonus points for showing the mode would be good (pipe, etc).

Reply:
Done